### ISSUE TITLE
Panic on OSX ("Invalid argument")

### ISSUE THREAD

**Issue by madadam** [maidsafe/crust#577]
The "send_and_receive" test (in https://github.com/maidsafe/crust/pull/570 here: https://github.com/madadam/crust/blob/stress-test/tests/send_and_receive.rs, not yet merged to master) fails on OSX with this stacktrace:

```
thread 'RaiiBootstrap' panicked at '

 =======================================================================================
| Result evaluated to Err: Error { repr: Os { code: 22, message: "Invalid argument" } } |
 =======================================================================================

', src/connection.rs:224
stack backtrace:
   1:        0x10224abb8 - sys::backtrace::tracing::imp::write::h7add7fdda97786a1s4u
   2:        0x10224ca95 - panicking::default_handler::_$u7b$$u7b$closure$u7d$$u7d$::closure.44030
   3:        0x10224c6b8 - panicking::default_handler::h72a4f6df9637bc80Rvz
   4:        0x10223beb6 - sys_common::unwind::begin_unwind_inner::hf798769ad97c097fe3t
   5:        0x10223c34e - sys_common::unwind::begin_unwind_fmt::hb2ee19b4aa59ba8ak2t
   6:        0x101ee0f2c - connection::connect_tcp_endpoint::_$u7b$$u7b$closure$u7d$$u7d$::closure.23153
   7:        0x101ee09bb - result::Result<T, E>::unwrap_or_else::h8718638085262797868
   8:        0x101eb09a6 - connection::connect_tcp_endpoint::hd5904597d9cef2e9nkb
   9:        0x101ebd30c - connection::connect::hd21b7049d845f2b8fjb
  10:        0x101eba75d - bootstrap::RaiiBootstrap::bootstrap::h93a93d7ac124546dT7a
  11:        0x101eb9dd8 - bootstrap::RaiiBootstrap::new::_$u7b$$u7b$closure$u7d$$u7d$::closure.22280
  12:        0x101eb9677 - std::thread::Builder::spawn::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::closure.22269
  13:        0x101eb95c8 - sys_common::unwind::try::try_fn::h3100308267746893306
  14:        0x10224a06b - __rust_try
  15:        0x102247463 - sys_common::unwind::inner_try::hdaa7ae415a983f2ag0t
  16:        0x101eb9505 - sys_common::unwind::try::h245024374647048752
  17:        0x101eb9349 - std::thread::Builder::spawn::_$u7b$$u7b$closure$u7d$$u7d$::closure.22266
  18:        0x101eba0e7 - boxed::F.FnBox<A>::call_box::h4038387312002249685
  19:        0x10224bd1c - sys::thread::Thread::new::thread_start::h45441f586afbe2dciKy
  20:     0x7fff855e4059 - _pthread_body
  21:     0x7fff855e3fd6 - _pthread_start
```

This suggest problem in crust master, as the PR only adds this test and does not change anything in crust itself.

Fixing this would allow merging https://github.com/maidsafe/crust/pull/570 and closing https://github.com/maidsafe/crust/issues/569.

**Comment by madadam** [comment]
New information:

- This error occurs no matter how many messages are sent. In fact, it occurs even when no messages are sent.
- It occurs even with as few as two clients, but only if they are spawned in parallel (each in its own thread). If spawned sequentially, it doesn't occur.
- The crash originates here: https://github.com/maidsafe/crust/blob/master/src/connection.rs#L220 - `peer_addr` sometimes returns error here.  
- This error occurs only when listening over TCP. UDP seems to be unaffected.

**Comment by madadam** [comment]
Fixed in #570 . Once it is merged, this issue can be closed.
